frontend/apps/reader/modules/readersearch.lua

```diff
@@ -205,44 +353,82 @@ function ReaderSearch:onShowSearchDialog(text, direction)
             UIManager:setDirty(self.dialog, "ui")
         end,
     }
-    do_search(self.searchFromCurrent, text, direction)()
-    UIManager:show(self.search_dialog)
-    --- @todo regional
-    UIManager:setDirty(self.dialog, "partial")
+    if regex and isSlowRegex(text) then
+        self.wait_button.alpha = nil
+        UIManager:show(self.wait_button)
+        UIManager:tickAfterNext(function()
+            do_search(self.searchFromCurrent, text, direction, regex, case_insensitive)()
+            UIManager:close(self.wait_button)
+            UIManager:show(self.search_dialog)
+            --- @todo regional
+            UIManager:setDirty(self.dialog, "partial")
+        end)
+    else
+        do_search(self.searchFromCurrent, text, direction, regex, case_insensitive)()
+        UIManager:show(self.search_dialog)
+        --- @todo regional
+        UIManager:setDirty(self.dialog, "partial")
+    end
+
     return true
 end
 
-function ReaderSearch:search(pattern, origin)
+-- if regex == true, use regular expression in pattern
+-- if case == true or nil, the search is case insensitive
+function ReaderSearch:search(pattern, origin, regex, case_insensitive)
     logger.dbg("search pattern", pattern)
     local direction = self.direction
-    local case = self.case_insensitive
     local page = self.view.state.page
-    return self.ui.document:findText(pattern, origin, direction, case, page)
+    if case_insensitive == nil then
+        case_insensitive = true
+    end
+    Device:setIgnoreInput(true)
+    local retval, words_found = self.ui.document:findText(pattern, origin, direction, case_insensitive, page, regex, self.max_hits)
+    Device:setIgnoreInput(false)
+    local regex_retval = self.ui.document:getAndClearRegexSearchError();
+    if regex_retval ~= 0 then
+        local error_message
+        if SRELL_ERROR_CODES[regex_retval] then
+            error_message = SRELL_ERROR_CODES[regex_retval]
+        else
+            error_message = _("Unspecified error")
+        end
+        UIManager:show(Notification:new{
+            text = error_message,
+            timeout = 4,
```

Comment:
What kind of error message we get at this point? Is a small notification at top visible for 4 second enough ? No risk for the user to need to read and analyzer the error message (so, something longer and not timeouting ?)

Reply:
This would be the message (or some unspecified internal srell exception, which I have never seen):
```
SRELL_ERROR_CODES[111] = _("Expression too complex, some hits will not be shown.")
```

It's difficult for me to decide what is long enough (as I am trained to look at the "right" position to detect regex behaviour).
Normally you don't expect a search to give an error message, so _no_ timeout might be a better solution.

Reply:
OK, so its just the equivalent of "too many hits" - the info about a regex being invalid is shown in an InfoMessage at text input type, where we have the time to read and fix it. So a notification is fine.